Modules/posixmodule.c

```diff
@@ -1652,56 +1589,59 @@ win32_xstat_impl(const wchar_t *path, struct _Py_stat_struct *result,
            If the latter, we can use attributes_from_dir. */
         DWORD lastError = GetLastError();
         if (lastError != ERROR_ACCESS_DENIED &&
-            lastError != ERROR_SHARING_VIOLATION)
+            lastError != ERROR_SHARING_VIOLATION &&
+            lastError != ERROR_INVALID_PARAMETER)
             return -1;
         /* Could not get attributes on open file. Fall back to
            reading the directory. */
         if (!attributes_from_dir(path, &info, &reparse_tag))
             /* Very strange. This should not fail now */
             return -1;
+        is_link = _Py_is_reparse_link(path, reparse_tag);
         if (info.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
-            if (traverse) {
+            if (!is_link || traverse) {
                 /* Should traverse, but could not open reparse point handle */
                 SetLastError(lastError);
                 return -1;
             }
         }
     } else {
-        if (!GetFileInformationByHandle(hFile, &info)) {
+        // Get file attributes + reparse tag first
+        if (!GetFileInformationByHandleEx(hFile, FileAttributeTagInfo,
+                                          &taginfo, sizeof(taginfo))) {
             CloseHandle(hFile);
```

Comment:
If querying the `FileAttributeTagInfo` fails, assume the filesystem doesn't support this info class. For example, VirtualBox shared folders use a VBoxSF file system that fails in this case, but that shouldn't fail the whole `stat` call.

Reply:
Thanks, it seems I dropped that code somewhere along the way. Pushed fix for this (according to original matching of `GetLastError()` as outlined previously), together with the cleanup of the other comment.